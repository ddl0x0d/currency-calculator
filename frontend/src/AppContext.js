import React from 'react';

const AppContext = React.createContext({ currencies: [] });

export default AppContext;
