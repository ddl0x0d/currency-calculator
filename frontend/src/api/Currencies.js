import axios from 'axios';
import { extractMessages } from './errorService';

const api = '/api/currencies';

export const get = async () => {
  try {
    const response = await axios.get(`${api}`);
    return response.data.data;
  } catch (e) {
    throw extractMessages(e);
  }
};

export const convert = async request => {
  try {
    const response = await axios.post(`${api}/convert`, request);
    return response.data.data.toFixed(2);
  } catch (e) {
    throw extractMessages(e);
  }
};

export default { get, convert };
