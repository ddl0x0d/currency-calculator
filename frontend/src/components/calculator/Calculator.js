import AmountInput from '../common/AmountInput';
import AppContext from '../../AppContext';
import Col from 'react-bootstrap/Col';
import CurrencySelector from '../common/CurrencySelector';
import Currencies from '../../api/Currencies';
import Row from 'react-bootstrap/Row';
import React, { useContext, useEffect, useState } from 'react';

export default function Calculator() {
  const [amount, setAmount] = useState(0);
  const [from, setFrom] = useState('');
  const [to, setTo] = useState('');
  const [conversion, setConversion] = useState('');
  const { showErrors } = useContext(AppContext);

  useEffect(() => {
    if (amount && amount > 0 && from && to && from !== to) {
      Currencies.convert({ amount, from, to })
        .then(conversion => {
          setConversion(`${conversion} ${to}`);
          showErrors([]);
        })
        .catch(errors => {
          setConversion('');
          showErrors(errors);
        });
    } else {
      setConversion('');
    }
  }, [amount, from, to]);

  return (
    <>
      <Row className="justify-content-center">
        <Col xs={3} md={2} xl={1}>
          <AmountInput
            id="amount"
            label="Amount"
            value={amount}
            handleChange={value => setAmount(value)}
          />
        </Col>
        <Col xs={3} md={2} xl={1}>
          <CurrencySelector
            id="from"
            label="From"
            value={from}
            handleChange={value => setFrom(value)}
          />
        </Col>
        <Col xs={3} md={2} xl={1}>
          <CurrencySelector
            id="to"
            label="To"
            value={to}
            handleChange={value => setTo(value)}
          />
        </Col>
      </Row>
      <Row className="mt-3">
        <Col>
          <h3 id="conversion" className="text-center font-weight-bolder">
            {conversion}
          </h3>
        </Col>
      </Row>
    </>
  );
}
