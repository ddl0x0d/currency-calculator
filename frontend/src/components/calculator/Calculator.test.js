import AppContext from '../../AppContext';
import Calculator from './Calculator';
import Currencies from '../../api/Currencies';
import { mount, shallow } from 'enzyme';
import React from 'react';

jest.mock('../../api/Currencies');

describe('<Calculator />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Calculator />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('convert currency', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    const showErrors = jest.fn().mockName('showErrors');
    const wrapper = mount(
      <AppContext.Provider value={{ currencies: [], showErrors }}>
        <Calculator />
      </AppContext.Provider>
    );

    const change = (selector, value) => {
      const element = wrapper.find(selector);
      element.simulate('change', { target: { value } });
    };

    beforeEach(() => {
      change('input#amount', 0);
      change('select#from', '');
      change('select#to', '');
    });

    it('succeeds', done => {
      Currencies.convert.mockName('convert').mockResolvedValue('85.00');
      change('input#amount', 100);
      change('select#from', 'EUR');
      change('select#to', 'GBP');

      setTimeout(() => {
        wrapper.update();

        const request = { amount: 100, from: 'EUR', to: 'GBP' };
        expect(Currencies.convert).toBeCalledWith(request);

        const conversion = wrapper.find('#conversion').text();
        expect(conversion).toBe('85.00 GBP');
        expect(showErrors).toBeCalledWith([]);
        done();
      }, 100);
    });

    it('fails', done => {
      const errors = [{ message: 'error' }];
      Currencies.convert.mockName('convert').mockRejectedValue(errors);
      change('input#amount', 100);
      change('select#from', 'EUR');
      change('select#to', 'GBP');

      setTimeout(() => {
        wrapper.update();

        const request = { amount: 100, from: 'EUR', to: 'GBP' };
        expect(Currencies.convert).toBeCalledWith(request);

        const conversion = wrapper.find('#conversion').text();
        expect(conversion).toBe('');
        expect(showErrors).toBeCalledWith(errors);
        done();
      }, 100);
    });
  });
});
