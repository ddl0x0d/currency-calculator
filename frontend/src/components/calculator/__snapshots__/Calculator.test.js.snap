// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> renders correctly 1`] = `
<Fragment>
  <Bootstrap(Row)
    className="justify-content-center"
  >
    <Col
      as="div"
      md={2}
      xl={1}
      xs={3}
    >
      <AmountInput
        handleChange={[Function]}
        id="amount"
        label="Amount"
        step={1}
        value={0}
      />
    </Col>
    <Col
      as="div"
      md={2}
      xl={1}
      xs={3}
    >
      <CurrencySelector
        handleChange={[Function]}
        id="from"
        label="From"
        value=""
      />
    </Col>
    <Col
      as="div"
      md={2}
      xl={1}
      xs={3}
    >
      <CurrencySelector
        handleChange={[Function]}
        id="to"
        label="To"
        value=""
      />
    </Col>
  </Bootstrap(Row)>
  <Bootstrap(Row)
    className="mt-3"
  >
    <Col
      as="div"
    >
      <h3
        className="text-center font-weight-bolder"
        id="conversion"
      />
    </Col>
  </Bootstrap(Row)>
</Fragment>
`;
